README.md
# Connecting to services

[![GitHub Release](https://img.shields.io/github/release/nhsuk/connecting-to-services.svg)](https://github.com/nhsuk/connecting-to-services/releases/latest/)
[![Greenkeeper badge](https://badges.greenkeeper.io/nhsuk/connecting-to-services.svg)](https://greenkeeper.io/)
[![Build Status](https://travis-ci.org/nhsuk/connecting-to-services.svg?branch=master)](https://travis-ci.org/nhsuk/connecting-to-services)
[![Coverage Status](https://coveralls.io/repos/github/nhsuk/connecting-to-services/badge.svg?branch=master)](https://coveralls.io/github/nhsuk/connecting-to-services?branch=master)

> A service to help people connect to pharmacy services.

## Installation

Clone the repo: `git clone https://github.com/nhsuk/connecting-to-services.git`
and review the [`scripts`](scripts) to get up and running.

## Changing the date and time

The application will display a message for each open pharmacy on a bank holiday.
In order to test this functionality the date can be set when the application
starts. In order to do this the environment variable, `DATETIME` needs to be
set. For example, if the date wanted to be to set to Christmas day 2017 the
following can be run on the command line
`DATETIME=2017-12-25T12:00:00 ./scripts/start`.

## Testing

The application uses [Docker](https://www.docker.com/) to run in containers.
Development is typically done on the host machine. Files are loaded into the
container and changes are automatically updated.

Use the `test` script for continuous testing during development.

## Test environments

As the application is being developed, every Pull Request has its own test
environment automatically built and deployed to.

Every environment apart from the one we want the public to access requires
basic authentication to access. The username and password are not secret, in
fact they are included within environment variable table below.
The intention with the authentication challenge is to prevent people whom may
stumble across the site and not realise it is for testing, it also prevents
access by search engines and other bots.

## BackstopJS tests

There are some BackstopJS UI tests located in [backstop-tests](backstop-tests).
These can be run in Docker using the following instructions:
* From the project root directory run `./scripts/backstop`
* An html report will be generated. It will be accessible at
  `./backstop-tests/backstop_data/html_report/index.html`

## Environment variables

Environment variables are expected to be managed by the environment in which
the application is being run. This is best practice as described by
[twelve-factor](https://12factor.net/config).

In order to protect the application from starting up without the required
env vars in place
[require-environment-variables](https://www.npmjs.com/package/require-environment-variables)
is used to check for the env vars that are required for the application to run
successfully.
This happens during the application start-up. If an env var is not found the
application will fail to start and an appropriate message will be displayed.

Environment variables are used to set application level settings for each
environment.

Environment variables need to be available within the process the application
is running in. This can be done by setting the env vars on the command line or
by adding them to a `.env` file. The `.env` file is excluded from being added
to source control, it is likely to contain sensitive data which should not be
checked in. A sample `.env` file is available `.env-sample` with some setting
that can be renamed to `.env` so the application will use this.

| Variable                           | Description                                                                                                                            | Default                                             | Required  |
| :--------------------------------- | :------------------------------------------------------------------------------------------------------------------------------------- | :----------------------------------------           | :-------- |
| `ADOBE_TRACKING_URL`               | [Adobe Analytics](https://www.adobe.com/analytics/adobe-analytics.html) Dynamic Tag Management URL                                     |                                                     | No        |
| `BACKSTOP_TEST_URL`                | The fully qualified domain against which the BackstopJS UI tests will be run e.g. `http://beta.nhs.uk`                                 | http://beta.nhs.uk                                  | Yes       |
| `BASIC_AUTH`                       | An MD5 encrypted [htpasswd](https://httpd.apache.org/docs/2.4/misc/password_encryptions.html)                                          | See value set in Vault for `BASIC_AUTH` in defaults |           |
| `BUSINESS_HOURS_START_HOUR`        | Business hours start hour                                                                                                              | 8                                                   | No        |
| `BUSINESS_HOURS_START_MINUTE`      | Business hours start minute                                                                                                            | 0                                                   | No        |
| `BUSINESS_HOURS_END_HOUR`          | Business hours end hour                                                                                                                | 18                                                  | No        |
| `BUSINESS_HOURS_END_MINUTE`        | Business hours end minute                                                                                                              | 0                                                   | No        |
| `COOKIEBOT_SCRIPT_URL`             | The URL for the in-house implementation of Cookiebot                                                                                   | //assets.nhs.uk/scripts/cookie-consent.js           |           |
| `DATETIME`                         | The initial display of results is determined by the time of the search. Bank holiday messaging is determined by the date of the search |                                                     | No        |
| `HOTJAR_ANALYTICS_TRACKING_ID`     | [Hotjar](https://www.hotjar.com/) tracking id                                                                                          |                                                     | No        |
| `LOG_LEVEL`                        | Numeric [log level](https://github.com/trentm/node-bunyan#levels)                                                                      | Depends on `NODE_ENV`                               |           |
| `NODE_ENV`                         | Node environment                                                                                                                       | development                                         |           |
| `PORT`                             | Server port                                                                                                                            | 3000                                                |           |
| `SEARCH_API_HOST`                  | Hostname for Organisation Search API                                                                                                   | api.nhs.uk                                          | No        |
| `SEARCH_API_KEY `                  | Access key for Organisation Search API                                                                                                 |                                                     | Yes       |
| `SEARCH_API_VERSION`               | Organisation Search API version number                                                                                                 | 1                                                   | No        |

## FAQ

* Is the application failing to start?
  * Ensure all expected environment variables are available within the environment
  * If set, `LOG_LEVEL` must be a number and one of the defined [log levels](https://github.com/trentm/node-bunyan#levels)
  * Check for messages in the logs

## Architecture Decision Records

This repo uses
[Architecture Decision Records](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions)
to record architectural decisions for this project.
They are stored in [doc/adr](doc/adr).


CHANGELOG.md
1.19.0 / TBC
===================
- Update npm dependencies
- Update basic auth creds
- Remove redundant link from search page breadcrumb
- Add bank holiday dates for 2020 & 2021
- Add ability to use .env files for environment variables

1.18.1 / 2019-12-05
===================
- Update Org Name to be the link to PIMS overview page tab
- Remove the services link

1.18.0 / 2019-11-14
===================
- Update URLs to `/service-search/find-a-pharmacy/`

1.17.3 / 2019-10-25
===================
- Update cookie link to be relative

1.17.2 / 2019-10-24
===================
- Update header search to use nhs.uk

1.17.1 / 2019-10-23
===================
- Update favicons
- Fix opening times styling bug
- Validation accessibility enhancements

1.17.0 / 2019-10-23
===================
- Add info json response for diagnostics

1.16.0 / 2019-10-23
===================
- Update project to use nhsuk frontend library

1.15.5 / 2019-10-20
===================
- Fix logic error in azure search

1.15.3 / 2019-10-10
===================
- Remove unused env vars from README
- Add subtype clause to pharmacy query
- Fix code coverage check

1.15.2 / 2019-10-09
===================
- Switch to using Azure Search for organisation queries
- Update npm dependencies
- Switch from istanbul to nyc for code coverage
- Add guard against inconsistency in opening time data 
- Remove duplicate address lines

1.14.0 / 2019-07-18
===================
- Use full path for Open Graph image
- Do not use capitalised convention for constant variables
- Update npm dependencies
- Upgrade Docker container to use latest node LTS version

1.13.0 / 2019-07-16
===================
- Remove the specific version of python from the Dockerfile
- Update npm dependencies
- Use node LTS version for travis
- Upgrade Docker container to use latest node LTS version
- Add additional (and update existing) metadata properties to meet minimum set
- Allow Adobe and Hotjar to connect from additional domains

1.12.0 / 2019-03-20
===================
- Update npm dependencies
- Upgrade Docker container to `node:10.15.3-alpine`
- Remove newrelic application monitoring
- Remove simple cookie banner
- Remove Webtrends (no longer used)
- Remove Google Analytics (no longer used)
- Remove Google Sitelinks Search Metadata

1.11.0 / 2018-11-29
===================
- Upgrade Docker container to `node:10.14.0-alpine`
- Update npm dependencies

1.10.0 / 2018-11-20
===================
- Use new frontend styles for opening times expander
- Make the opening times expander full width on screens below tablet
- Update CSP to allow Adobe Analytics resources to run
- Create `digitalData` object for use by Adobe Analytics
- Update npm dependencies

1.9.0 / 2018-10-09
===================
- Update npm dependencies
- Accessibility improvements
- Upgrade to `node:8.12.0-alpine`
- Prevent IE 11 from focusing on SVGs
- Fix skip to content for voice over

1.8.0 / 2018-09-14
===================
- Update npm dependencies
- Add Adobe Analytics

1.7.0 / 2018-09-07
===================
- Remove unnecessary npm dependencies
- Do not link to beta home page
- Update npm dependencies
- Update 'contact-us' footer link
- Update 'care and support' header link

1.6.0 / 2018-08-30
===================
- Update footer links to new pages
- Update cookie banner link to cookie policy
- Update favicon with latest design
- Remove Heatmaps from Webtrends (the feature has been discontinued)
- Upgrade Docker container to `node:8.11.4-alpine`
- Update python used in container
- Update npm dependencies
- Anonymise IP used by Webtrends

1.5.0 / 2018-08-15
===================
- Update npm dependencies
- Remove references to `NHS.UK` in favour of `NHS`
- Set background to blue
- Fix input width
- Update header search placeholder text
- Remove aria-labelledby on the aria-hidden NHS logo

1.4.0 / 2018-08-09
===================
- Update npm dependencies
- Update code for latest `eslint-config-airbnb-base` rules
- Update for site redesign
- Update IE css to new design
- Update print css to new design

1.3.0 / Changes reverted - not released
======================================
- Update Webtrends and GA configuration for GDPR compliance
- Replace bespoke cookie banner with Cookiebot
- Only enable GA, Webtrends and Hotjar if consented from Cookiebot

1.2.0 / 2018-05-03
===================
- Use first non-empty address line in screen reader only text for Map and Direction and Services Available links
- Allow fonts from hotjar.com in Content Security Policy

1.1.0 / 2018-04-17
===================
- Update npm dependencies
- Add links to internet services & appliance contractors
- Upgrade Docker container to `node:8.11.1-alpine`

1.0.0 / 2018-04-04
===================
- Remove snyk hooks
- Remove `This is new banner`
- Update npm dependencies
- Changed open on/off toggle to text and link only
- Skip to page content links & accessibility help pages
- Stopped using details summary for show/hide panels due to lack of IE & accessibility support
- General accessibility improvements

0.37.0 / 2018-03-23
===================
- Upgrade to `postcodesio-client@0.3.3`
- Upgrade to `eslint-config-nhsuk@0.13.0` and apply fixes
- Display open results by default for times deemed out of business hours
- Update npm dependencies

0.36.0 / 2018-03-14
===================
- Add newrelic application monitoring
- Update npm dependencies
- Upgrade Docker container to `node:8.10.0-alpine`

0.35.1 / 2018-03-02
===================
- Revert version of performance test script

0.35.0 / N/A
============
- Run performance test with basic auth

0.34.1 / N/A
============
- Responsive design and spacing fixes

0.34.0 / N/A
============
- Upgrade python in Alpine to 2.7.14-r0
- Add on/off toggle for showing open/closed pharmacies

0.33.0 / 2018-02-22
===================
- Add hidden assistive text to meet accessibility standards
- IE bug fix for opening times menu toggle

0.32.0 / 2018-02-20
===================

config/config.js
const path = require('path');

const rootPath = path.normalize(`${__dirname}/..`);

module.exports = {
  analytics: {
    adobeTrackingUrl: process.env.ADOBE_TRACKING_URL,
    hotjarId: process.env.HOTJAR_ANALYTICS_TRACKING_ID,
  },
  app: {
    name: 'connecting-to-services',
  },
  businessHours: {
    end: {
      hour: process.env.BUSINESS_HOURS_END_HOUR || 18,
      minute: process.env.BUSINESS_HOURS_END_MINUTE || 0,
    },
    start: {
      hour: process.env.BUSINESS_HOURS_START_HOUR || 8,
      minute: process.env.BUSINESS_HOURS_START_MINUTE || 0,
    },
  },
  cookiebot: {
    scriptUrl: process.env.COOKIEBOT_SCRIPT_URL || '//assets.nhs.uk/scripts/cookie-consent.js',
  },
  dotenvPath: process.env.DOTENV_PATH || '.env',
  env: process.env.NODE_ENV || 'development',
  port: process.env.PORT || 3000,
  root: rootPath,
  search: {
    apiKey: process.env.SEARCH_API_KEY,
    host: process.env.SEARCH_API_HOST || 'api.nhs.uk',
    maxNumberOfResults: process.env.SEARCH_MAX_RESULTS || 10,
    version: process.env.SEARCH_API_VERSION || '1',
  },
  timezone: 'Europe/London',
};
